### ISSUE TITLE
Type of a variable changes before new value is assigned

### ISSUE THREAD

**Issue by antonilol** [microsoft/TypeScript#52657]
# Bug Report

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.5#code/N4WAUABFEDYKYBcIIFwQHYFcC2AjOATgNoC6EAPhAM4IECW6A5qeONMhALwRECsJrSNAD0w5Gix5CLMGxFiA7gHsCAaypyoSbggB02AIYAHABTouAPgy6ESgGJ0AHnAAmJgMwBKTwG5wAX0FQISh4JFQMHHxiMkoaeiYZTQ5uPgFZEIhRcUipGMF2bJclOCp0AHIkZTUIfABjA0wqOA46gAsDJldkJVzo0gpqWgZmdPZtZH1jExMAfQAaCDpPS2QiOhIbeydXD28-MECwIA)

### 💻 Code

```ts
let t: number[] | string[]

t = [5]

// t: number[]

// doesn't work because t changed to number[] | string[]
t = t.map((_, i) => t[i].toFixed(3));
```

### 🙁 Actual behavior

`t` is a `number[] | string[]` inside the map function, which is executed before t is assigned the new array of strings

### 🙂 Expected behavior

`t` stays `number[]` inside the map function, and changes to `string[]` after the `string[]` returned by `Array.map` gets assigned to it

**Comment by fatcerberus** [comment]
Yet another duplicate of #9998.  FWIW there’s an issue template explaining this (interestingly this report covers both cases at the same time).

**Comment by antonilol** [comment]
> Yet another duplicate of #9998.

oops, missed that one, closing this issue then

> FWIW there’s an issue template explaining this (“types not correct in/with callback”).

i see it know you told me, but when creating an issue the template is chosen based on what type of issue it is (in this case bug), i would not expect a message to avoid duplicates there

**Comment by fatcerberus** [comment]
Yeah, running into #9998 is like a rite of passage for TS developers. 😅   That’s the only reason there’s that special template for dupes, because basically *everyone* hits this sooner or later and it tends to look like a bug at first.  It would be nice if eventually TS could be smarter about this but it’s not a trivial fix.
